src/libstd/sys/unix/fs.rs

```diff
@@ -732,17 +791,126 @@ pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
-    if !from.is_file() {
+    use fs::OpenOptions;
+    let mut from_opts = OpenOptions::new();
+    let mut reader = try!(from_opts.read(true).open(&from));
+    let metadata = try!(reader.metadata());
+    let attr = metadata.as_inner();
+    if attr.file_type().is_file() == false {
         return Err(Error::new(ErrorKind::InvalidInput,
                               "the source path is not an existing regular file"))
     }
 
-    let mut reader = try!(File::open(from));
-    let mut writer = try!(File::create(to));
-    let perm = try!(reader.metadata()).permissions();
+    let mut to_opts = OpenOptions::new();
+    let mut writer = try!(to_opts.write(true)
+                                 .create(true)
+                                 .truncate(true)
+                                 .mode(attr.perm().mode())
+                                 .open(&to));
 
     let ret = try!(io::copy(&mut reader, &mut writer));
-    try!(set_permissions(to, perm));
+    try!(writer.as_inner().set_times(try!(attr.accessed()),
+                                     try!(attr.modified()),
+                                     attr.created().ok()
+    ));
     Ok(ret)
 }
+
+#[cfg(test)]
+mod tests {
+    use prelude::v1::*;
+    use env;
+    use fs;
+    use path::PathBuf;
+    use rand::{self, Rng};
+    use time::Duration;
+
+    macro_rules! check { ($e:expr) => (
+        match $e {
+            Ok(t) => t,
+            Err(e) => panic!("{} failed with: {}", stringify!($e), e),
+        }
+    ) }
+
+    macro_rules! error { ($e:expr, $s:expr) => (
+        match $e {
+            Ok(_) => panic!("Unexpected success. Should've been: {:?}", $s),
+            Err(ref err) => assert!(err.to_string().contains($s),
+                                    format!("`{}` did not contain `{}`", err, $s))
+        }
+    ) }
+
+    // from std::time, but with a resolution of 1 second.
+    macro_rules! assert_almost_eq {
+        ($a:expr, $b:expr) => ({
+            let (a, b) = ($a, $b);
+            if a != b {
+                let (a, b) = if a > b {(a, b)} else {(b, a)};
+                assert!(a.sub_duration(&Duration::new(1, 0)) <= b);
+            }
+        })
+    }
+
+    pub struct TempDir(PathBuf);
+
+    impl TempDir {
+        fn join(&self, path: &str) -> PathBuf {
+            let TempDir(ref p) = *self;
+            p.join(path)
+        }
+    }
+
+    impl Drop for TempDir {
+        fn drop(&mut self) {
+            // Gee, seeing how we're testing the fs module I sure hope that we
+            // at least implement this correctly!
+            let TempDir(ref p) = *self;
+            check!(fs::remove_dir_all(p));
+        }
+    }
+
+    pub fn tmpdir() -> TempDir {
+        let p = env::temp_dir();
+        let mut r = rand::thread_rng();
+        let ret = p.join(&format!("rust-{}", r.next_u32()));
+        check!(fs::create_dir(&ret));
+        TempDir(ret)
+    }
+
+    #[test]
+    fn copy_preserves_times() {
```

Comment:
I wonder if this should be a cross-platform test? Windows should also copy file times, right?